src/go/rdctl/pkg/snapshot/copyFile_darwin.go

```diff
@@ -0,0 +1,43 @@
+package snapshot
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/sys/unix"
+)
+
+// Copies a file from src to dst. If copyOnWrite is true, attempts to
+// use clonefile syscall to do the copy. If clonefile is not supported
+// by the underlying filesystem, or src and dst are on different
+// drives, falls back to a plain copy. If copyOnWrite is false, does a
+// plain copy.
+func copyFile(dst, src string, copyOnWrite bool, fileMode os.FileMode) error {
```

Comment:
I'm wondering if you can fold the darwin `copyFile` and linux `copyFile` into the same code, with a platform-specific callback function to invoke if `copyOnWrite` is true.

It's hard to tell, because the two functions have similar bits, but they're in different order:

`copyFile_darwin.go`: do `os.MkdirAll(...)` first, later `os.Open(src)`.
`copyFile_linux.go`: do `os.Open(src)` first, later `os.MkdirAll(...)`

It could be that these could be done in parallel, but I'd rather have them done in one function in either
order. Something like:

`copyFile_darwin.go`: 
```
func copyFile(...) {
   return commonCopyFile(..., func() error {
                // darwin callback if `copyOnWrite` is true
		if err := unix.IoctlFileClone(int(dstFd.Fd()), int(srcFd.Fd())); err == nil {
			return nil
		} else if !errors.Is(err, unix.ENOTSUP) {
			return fmt.Errorf("failed to ioctl_ficlone file: %w", err)
		}
})
```

Reply:
Point 1: regarding the use of build constraints, this would be awkward once we bring Windows into the picture. Right now we use `_linux.go` and `_darwin.go` to specify which functions are used on which platforms. If we make `copyFile` cross-platform, how do we handle things like file mode on Windows? Or the fact we aren't (I think) using copy-on-write syscalls on Windows? Do we introduce a separate function for Windows? In that case we would want to use the `_windows.go` and `_unix.go` build constraints. But `_unix.go` is only available in go 1.19+ (we are using 1.18 for `rdctl` at the moment). We could also decide which function to use at runtime, but I'd rather not do that.

Point 2: there are three things that led me to the current design: 
- the Linux `unix.IoctlFileClone` function takes file descriptors as arguments
- the macOS `unix.Clonefile` function takes paths as arguments
- copying a file without the use of the copy-on-write functions requires file descriptors

While I am conscious that we want to avoid premature optimization, it doesn't make much sense to (in the Linux case) open two files within the platform-specific copy-on-write function and then (if it fails in an anticipated way) open those files again for the regular copy in the cross-platform copy function. Though this would make more sense in the case of macOS.

> copyFile_darwin.go: do os.MkdirAll(...) first, later os.Open(src).
> copyFile_linux.go: do os.Open(src) first, later os.MkdirAll(...)

This is deliberate: for macOS the destination file's directory must exist before doing the `clonefile`, otherwise it fails. So we call `os.MkdirAll` before it. Whereas for Linux, I wanted to hold off on creating the directory until it was needed (i.e. right before the destination file was opened) so that it isn't left hanging there if we ran into an error opening the source file.

Basically there is a bunch of nuanced stuff going on. I had it de-duplicated (in a different way) before and changed it to what it is because it's simpler to understand and work on when it's one function per platform and flat.

Reply:
Great explanation. Maybe put a sentence before each of the two functions to explain the differences.